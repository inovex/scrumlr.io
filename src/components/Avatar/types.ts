export const AVATAR_SKIN_COLORS = ["Tanned", "Yellow", "Pale", "Light", "Brown", "DarkBrown", "Black"] as const;
export type AvatarSkinColor = (typeof AVATAR_SKIN_COLORS)[number];

export const AVATAR_TOP_TYPES = [
  "Eyepatch",
  "Hat",
  "Hijab",
  "LongHairBigHair",
  "LongHairBob",
  "LongHairBun",
  "LongHairCurly",
  "LongHairCurvy",
  "LongHairDreads",
  "LongHairFrida",
  "LongHairFro",
  "LongHairFroBand",
  "LongHairMiaWallace",
  "LongHairNotTooLong",
  "LongHairShavedSides",
  "LongHairStraight",
  "LongHairStraight2",
  "LongHairStraightStrand",
  "NoHair",
  "ShortHairDreads01",
  "ShortHairDreads02",
  "ShortHairFrizzle",
  "ShortHairShaggyMullet",
  "ShortHairShortCurly",
  "ShortHairShortFlat",
  "ShortHairShortRound",
  "ShortHairShortWaved",
  "ShortHairSides",
  "ShortHairTheCaesar",
  "ShortHairTheCaesarSidePart",
  "Turban",
  "WinterHat1",
  "WinterHat2",
  "WinterHat3",
  "WinterHat4",
] as const;
export type AvatarTopType = (typeof AVATAR_TOP_TYPES)[number];

export const AVATAR_CLOTHE_COLORS = [
  "Black",
  "Blue01",
  "Blue02",
  "Blue03",
  "Gray01",
  "Gray02",
  "Heather",
  "PastelBlue",
  "PastelGreen",
  "PastelOrange",
  "PastelRed",
  "PastelYellow",
  "Pink",
  "Red",
  "White",
] as const;
export type AvatarClotheColor = (typeof AVATAR_CLOTHE_COLORS)[number];

export const AVATAR_GRAPHIC_TYPES = ["Bat", "Cumbia", "Deer", "Diamond", "Hola", "Pizza", "Resist", "Selena", "Bear", "SkullOutline", "Skull"] as const;
export type AvatarGraphicType = (typeof AVATAR_GRAPHIC_TYPES)[number];

export const AVATAR_CLOTHE_TYPES = ["BlazerShirt", "BlazerSweater", "CollarSweater", "GraphicShirt", "Hoodie", "Overall", "ShirtCrewNeck", "ShirtScoopNeck", "ShirtVNeck"] as const;
export type AvatarClotheType = (typeof AVATAR_CLOTHE_TYPES)[number];

export const AVATAR_HAIR_COLORS = ["Auburn", "Black", "Blonde", "BlondeGolden", "Brown", "BrownDark", "PastelPink", "Blue", "Platinum", "Red", "SilverGray"] as const;
export type AvatarHairColor = (typeof AVATAR_HAIR_COLORS)[number];

export const AVATAR_FACIAL_HAIR_COLORS = ["Auburn", "Black", "Blonde", "BlondeGolden", "Brown", "BrownDark", "Platinum", "Red"] as const;
export type AvatarFacialHairColor = (typeof AVATAR_FACIAL_HAIR_COLORS)[number];

export const AVATAR_FACIAL_HAIR_TYPES = ["Blank", "BeardMedium", "BeardLight", "BeardMajestic", "MoustacheFancy", "MoustacheMagnum"] as const;
export type AvatarFacialHairType = (typeof AVATAR_FACIAL_HAIR_TYPES)[number];

export const AVATAR_ACCESSORIES_TYPES = ["Blank", "Kurt", "Prescription01", "Prescription02", "Round", "Sunglasses", "Wayfarers"] as const;
export type AvatarAccessoriesType = (typeof AVATAR_ACCESSORIES_TYPES)[number];

export const AVATAR_EYE_TYPES = ["Close", "Cry", "Default", "Dizzy", "EyeRoll", "Happy", "Hearts", "Side", "Squint", "Surprised", "Wink", "WinkWacky"] as const;
export type AvatarEyeType = (typeof AVATAR_EYE_TYPES)[number];

export const AVATAR_EYEBROW_TYPES = [
  "Angry",
  "AngryNatural",
  "Default",
  "DefaultNatural",
  "FlatNatural",
  "FrownNatural",
  "RaisedExcited",
  "RaisedExcitedNatural",
  "SadConcerned",
  "SadConcernedNatural",
  "UnibrowNatural",
  "UpDown",
  "UpDownNatural",
] as const;
export type AvatarEyebrowType = (typeof AVATAR_EYEBROW_TYPES)[number];

export const AVATAR_MOUTH_TYPES = ["Concerned", "Default", "Disbelief", "Eating", "Grimace", "Sad", "ScreamOpen", "Serious", "Smile", "Tongue", "Twinkle", "Vomit"] as const;
export type AvatarMouthType = (typeof AVATAR_MOUTH_TYPES)[number];
