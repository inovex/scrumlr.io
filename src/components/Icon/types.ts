export type IconNames =
  | 'phase1-create'
  | 'phase1-communicate'
  | 'phase1-combine'
  | 'phase2-vote'
  | 'phase3-discuss'
  | 'phase3-define'
  | 'circle-arrow-left'
  | 'circle-arrow-right'
  | 'circle-add'
  | 'circle'
  | 'more'
  | 'plus'
  | 'circle-selection'
  | 'circle-selection-grey'
  | 'logout'
  | 'download'
  | 'edit'
  | 'donate'
  | 'pencil'
  | 'focus'
  | 'close'
  | 'close20'
  | 'close-circle'
  | 'feedback'
  | 'key-up'
  | 'key-down'
  | 'key-esc'
  | 'settings'
  | 'board'
  | 'trash'
  | 'chevron-left'
  | 'chevron-right'
  | 'stack-top'
  | 'stack-mid'
  | 'stack-hover'
  | 'arrow-left'
  | 'arrow-right';
