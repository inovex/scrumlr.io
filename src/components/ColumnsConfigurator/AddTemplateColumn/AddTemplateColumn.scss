@import "constants/style";

$add-column-width: 144px;
$add-column-height: 326px;

.add-template-column {
  display: flex;
  justify-content: center;
  align-items: center;

  min-width: $add-column-width;
  width: $add-column-width;

  height: $add-column-height;

  border: 3px dashed var(--accent-color--light);

  &--left {
    border-right: none;
    border-top-left-radius: $rounded--default;
    border-bottom-left-radius: $rounded--default;
  }
  &--right {
    border-left: none;
    border-top-right-radius: $rounded--default;
    border-bottom-right-radius: $rounded--default;
  }
}

.add-template-column__button {
  all: unset;

  &:enabled:active {
    .add-template-column__icon #circle {
      fill-opacity: 0.2;
    }
  }
}

.add-template-column__icon {
  width: 56px;
  height: 56px;

  color: var(--accent-color--light);

  &--disabled {
    color: $gray--800;
  }

  &:not(&--disabled) {
    cursor: pointer;

    &:hover {
      filter: drop-shadow(0 4px 14px rgba(var(--accent-color--light-rgb), 0.26));
    }
  }
}

[theme="dark"] {
  .add-template-column {
    border-color: var(--accent-color--dark);
  }

  .add-template-column__icon {
    color: var(--accent-color--dark);

    &:not(&--disabled):hover {
      filter: drop-shadow(0 4px 14px rgba(var(--accent-color--dark-rgb), 0.26));
    }
  }
}
