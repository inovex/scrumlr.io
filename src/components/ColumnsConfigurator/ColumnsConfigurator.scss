@import "constants/style";

.columns-configurator {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-shrink: 0;
  flex-grow: 1;

  // allow horizontal scroll
  overflow: auto hidden;
  width: 100%;
  max-width: 100%;
  height: 100%;

  @include scrollbar();
}

.columns-configurator__columns-wrapper {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-shrink: 0;
  flex-grow: 1;
}
