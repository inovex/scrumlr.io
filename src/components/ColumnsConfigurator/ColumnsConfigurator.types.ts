export type ColumnPlacement = "first" | "center" | "last" | "all";

export type AddTemplateColumnAlignment = "left" | "right";
