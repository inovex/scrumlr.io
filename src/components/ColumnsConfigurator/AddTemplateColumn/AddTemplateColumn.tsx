import {Color, getColorClassName} from "constants/colors";
import {ReactComponent as AddColumnIcon} from "assets/icons/add-column.svg";
import classNames from "classnames";
import {AddTemplateColumnAlignment} from "../ColumnsConfigurator.types";
import "./AddTemplateColumn.scss";

type AddTemplateColumnProps = {
  className?: string;
  alignment: AddTemplateColumnAlignment;
  color: Color;
  onClick: (alignment: AddTemplateColumnAlignment, color: Color) => void;
  disabled?: boolean;
};

export const AddTemplateColumn = (props: AddTemplateColumnProps) => (
  <div
    className={classNames(
      props.className,
      "add-template-column",
      `add-template-column--${props.alignment}`,
      {"add-template-column--disabled": props.disabled},
      getColorClassName(props.color)
    )}
  >
    <button className="add-template-column__button" onClick={() => props.onClick(props.alignment, props.color)} disabled={props.disabled}>
      <AddColumnIcon className={classNames("add-template-column__icon", {"add-template-column__icon--disabled": props.disabled})} />
    </button>
  </div>
);
