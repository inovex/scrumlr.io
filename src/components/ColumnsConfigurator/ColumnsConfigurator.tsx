import {uniqueId} from "underscore";
import classNames from "classnames";
import {useState} from "react";
import {EditableTemplateColumn, TemplateColumn} from "store/features";
import {Color, getColorClassName, getNextColor, getPreviousColor} from "constants/colors";
import {closestCenter, DndContext, DragEndEvent, DragOverEvent, DragOverlay, DragStartEvent, PointerSensor, useSensor, useSensors} from "@dnd-kit/core";
import {horizontalListSortingStrategy, SortableContext} from "@dnd-kit/sortable";
import {ColumnsConfiguratorColumn} from "./ColumnsConfiguratorColumn/ColumnsConfiguratorColumn";
import {AddTemplateColumn} from "./AddTemplateColumn/AddTemplateColumn";
import {calcPlacement} from "./ColumnsConfigurator.utils";
import {AddTemplateColumnAlignment} from "./ColumnsConfigurator.types";
import "./ColumnsConfigurator.scss";

export type ColumnsConfiguratorProps = {
  className: string;
  templateId: string;
  columns: EditableTemplateColumn[]; // without deleted cols
  addColumn: (templateColumn: TemplateColumn, index: number) => void;
  moveColumn: (fromIndex: number, toIndex: number) => void;
  editColumn: (templateColumn: EditableTemplateColumn, overwrite: Partial<EditableTemplateColumn>) => void;
  deleteColumn: (templateColumn: EditableTemplateColumn) => void;
};

export const ColumnsConfigurator = (props: ColumnsConfiguratorProps) => {
  // id of column which is actively being dragged
  const [activeElementId, setActiveElementId] = useState<string | null>(null);
  const [activeColumn, setActiveColumn] = useState<EditableTemplateColumn | null>(null);
  const [dropElementId, setDropElementId] = useState<string | null>(null);

  const sensors = useSensors(useSensor(PointerSensor));

  const handleDragStart = (event: DragStartEvent) => {
    const activeId = event.active.id as string;
    setActiveElementId(activeId);
    setActiveColumn(props.columns.find((c) => c.id === activeId)!);
  };

  const handleDragOver = (event: DragOverEvent) => {
    setDropElementId((event.over?.id as string) ?? null);
  };

  const handleDragEnd = (event: DragEndEvent) => {
    const {active, over} = event;

    if (active.id !== over?.id) {
      const oldIndex = props.columns.findIndex((c) => c.id === active.id);
      const newIndex = props.columns.findIndex((c) => c.id === over?.id);

      props.moveColumn(oldIndex, newIndex);
    }

    // don't forget to clear states
    setActiveElementId(null);
    setActiveColumn(null);
    setDropElementId(null);
  };

  // function to calculate where each column would be positioned if the drag-and-drop operation were completed at its current state
  const getPotentialIndex = (index: number) => {
    // no valid drag operation occurring
    if (!activeElementId || !dropElementId) return index;

    const dragElementIndex = props.columns.findIndex((c) => c.id === activeElementId);
    const dropElementIndex = props.columns.findIndex((c) => c.id === dropElementId);

    // if current index matches active index of drag element, the column is being dragged there,
    // so the potential index is that of the drop element
    if (index === dragElementIndex) {
      return dropElementIndex;
    }

    // dragging right: active index less the drop zone, and element is between them, the element shifts to the left
    if (dragElementIndex < dropElementIndex && index > dragElementIndex && index <= dropElementIndex) {
      return index - 1;
    }

    // dragging left: analogue to above, element shifts to the right
    if (dragElementIndex > dropElementIndex && index < dragElementIndex && index >= dropElementIndex) {
      return index + 1;
    }

    // otherwise the index remains unchanged
    return index;
  };

  // renders the drag overlay, which is the element which is dragged
  const renderDragOverlay = () => {
    if (!activeElementId || !activeColumn) return null;
    const activeColumnIndex = getPotentialIndex(props.columns.findIndex((c) => c.id === activeColumn.id));

    return (
      <ColumnsConfiguratorColumn
        className={classNames("columns-configurator__column", "columns-configurator__column--ghost")}
        column={activeColumn}
        // here, use actual index unlike potential index in the main part
        index={activeColumnIndex}
        placement={calcPlacement(activeColumnIndex, props.columns)}
        allColumns={props.columns}
      />
    );
  };

  const addTemplateColumn = (alignment: AddTemplateColumnAlignment, color: Color) => {
    const newColumn: TemplateColumn = {
      id: uniqueId("col"),
      template: props.templateId,
      name: `Column ${props.columns.length + 1}`,
      description: "",
      color,
      visible: true,
      index: -1, // will be overwritten anyway
    };

    // target index for the right has to take deleted cols which are still part of the array into account,
    // which is why we use the total col amount as new index
    const index = alignment === "left" ? 0 : props.columns.length;

    props.addColumn(newColumn, index);
  };

  if (props.columns.length === 0) {
    return null;
  }

  return (
    <DndContext sensors={sensors} collisionDetection={closestCenter} onDragStart={handleDragStart} onDragOver={handleDragOver} onDragEnd={handleDragEnd}>
      <SortableContext items={props.columns} strategy={horizontalListSortingStrategy}>
        <div className={classNames(props.className, "columns-configurator", getColorClassName("backlog-blue"))}>
          <AddTemplateColumn alignment="left" color={getPreviousColor(props.columns[0].color)} onClick={addTemplateColumn} />
          <div className="columns-configurator__columns-wrapper">
            {props.columns.map((column, index) => {
              const potentialIndex = getPotentialIndex(index);
              return (
                <ColumnsConfiguratorColumn
                  className="columns-configurator__column"
                  key={column.id}
                  column={column}
                  index={potentialIndex}
                  activeDrag={column.id === activeElementId}
                  activeDrop={column.id === dropElementId}
                  placement={calcPlacement(potentialIndex, props.columns)}
                  allColumns={props.columns}
                  editColumn={props.editColumn}
                  deleteColumn={props.deleteColumn}
                />
              );
            })}
          </div>
          <AddTemplateColumn alignment="right" color={getNextColor(props.columns[props.columns.length - 1].color)} onClick={addTemplateColumn} />
        </div>
      </SortableContext>

      <DragOverlay>{renderDragOverlay()}</DragOverlay>
    </DndContext>
  );
};
