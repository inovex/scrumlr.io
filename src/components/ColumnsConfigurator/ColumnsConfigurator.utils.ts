import {EditableTemplateColumn, ExistenceStatus, Template, TemplateColumn, TemplateWithColumns} from "store/features";
import {ColumnPlacement} from "./ColumnsConfigurator.types";

export const calcPlacement = (index: number, columns: TemplateColumn[]): ColumnPlacement => {
  const {length} = columns;
  if (length === 1) {
    return "all";
  }
  if (index === 0) {
    return "first";
  }
  if (index === length - 1) {
    return "last";
  }
  return "center";
};

// convert TemplateColumn to EditableColumn, used for test purposes
export const convertToEditableColumn = (c: TemplateColumn, overrideExistenceStatus?: Partial<ExistenceStatus>): EditableTemplateColumn =>
  ({
    ...c,
    persisted: false,
    mode: undefined,
    ...overrideExistenceStatus,
  }) as EditableTemplateColumn;

// from an array of all templates and column, create a TemplateAndColumns array with all corresponding pairs
export const mergeTemplateAndColumns = (templates: Template[], templateColumns: TemplateColumn[]) =>
  templates.map((template) => ({
    template,
    columns: templateColumns.filter((column) => column.template === template.id),
  })) as TemplateWithColumns[];

export const getTemplateAndColumnsByTemplateId = (templatesAndColumns: {templates: Template[]; templateColumns: TemplateColumn[]}, templateId: string) =>
  mergeTemplateAndColumns(templatesAndColumns.templates, templatesAndColumns.templateColumns).find((tac) => tac.template.id === templateId);
