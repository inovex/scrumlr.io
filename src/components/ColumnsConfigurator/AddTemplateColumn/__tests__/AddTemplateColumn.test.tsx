import {render} from "testUtils";
import {AddTemplateColumn} from "../AddTemplateColumn";
import {AddTemplateColumnAlignment} from "components/ColumnsConfigurator/ColumnsConfigurator.types";
import {Color} from "constants/colors";
import {fireEvent} from "@testing-library/react";

const renderAddTemplateColumn = (alignment: AddTemplateColumnAlignment, color: Color, disabled: boolean, onClick: (alignment: AddTemplateColumnAlignment, color: Color) => void) =>
  render(<AddTemplateColumn alignment={alignment} color={color} disabled={disabled} onClick={onClick} />);

describe("AddTemplateColumn", () => {
  it("should render correctly", () => {
    const {container} = renderAddTemplateColumn("left", "backlog-blue", false, jest.fn);
    expect(container).toMatchSnapshot();
  });

  it("should have no right border with left alignment", () => {
    const {container} = renderAddTemplateColumn("left", "backlog-blue", false, jest.fn);
    expect(container).toHaveStyle({"border-right": "none"});
  });

  it("should have no left border with right alignment", () => {
    const {container} = renderAddTemplateColumn("right", "backlog-blue", false, jest.fn);
    expect(container).toHaveStyle({"border-left": "none"});
  });

  it("should fire onClick fn parametrized with its alignment and color", () => {
    const alignment: AddTemplateColumnAlignment = "left";
    const color: Color = "poker-purple";
    const spy: jest.Mock<(alignment: AddTemplateColumnAlignment, color: Color) => void> = jest.fn();

    const {container} = renderAddTemplateColumn(alignment, color, false, spy);
    const addTemplateColumnButton = container.getElementsByClassName("add-template-column__button")[0];

    fireEvent.click(addTemplateColumnButton);
    expect(spy).toHaveBeenCalledWith(alignment, color);
  });

  it("should not fire onClick if disabled", () => {
    const spy = jest.fn();

    const {container} = renderAddTemplateColumn("right", "backlog-blue", true, spy);
    const addTemplateColumnButton = container.getElementsByClassName("add-template-column__button")[0];

    fireEvent.click(addTemplateColumnButton);
    expect(spy).not.toHaveBeenCalled();
  });
});
