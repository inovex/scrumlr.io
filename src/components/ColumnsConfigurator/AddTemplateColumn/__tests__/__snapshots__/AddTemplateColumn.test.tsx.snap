// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTemplateColumn should render correctly 1`] = `
<div>
  <div
    class="add-template-column add-template-column--left accent-color__backlog-blue"
  >
    <button
      class="add-template-column__button"
    >
      <svg
        class="add-template-column__icon"
      >
        add-column.svg
      </svg>
    </button>
  </div>
</div>
`;
